sa/sa.go

```diff
@@ -1771,3 +1772,39 @@ func addKeyHash(db db.Inserter, cert *x509.Certificate) error {
 	}
 	return db.Insert(khm)
 }
+
+// AddBlockedKey adds a key hash to the blockedKeys table
+func (ssa *SQLStorageAuthority) AddBlockedKey(ctx context.Context, req *sapb.AddBlockedKeyRequest) (*corepb.Empty, error) {
+	if req == nil || req.KeyHash == nil || req.Added == nil || req.Source == nil {
+		return nil, errIncompleteRequest
+	}
+	err := ssa.dbMap.Insert(&blockedKeyModel{
+		KeyHash: req.KeyHash,
+		Added:   time.Unix(0, *req.Added),
+		Source:  *req.Source,
+		Comment: req.Comment,
+	})
+	if err != nil {
+		if db.IsDuplicate(err) {
+			return nil, berrors.DuplicateError("cannot add a duplicate key")
```

Comment:
This would result in revocation requests returning an error to the end user. I think we should treat duplicate requests to insert the same key in the blocked keys list as non-errors: the key is already blocked, so the request is satisfied.

Reply:
This matches current behavior, if the certificate is already revoked, we fail out, rather than telling the user we succeeded.

Reply:
Yes, but this is actually different: If there are certificates A and B both with key K, once you revoke certificate A you could no longer revoke certificate B. You'd have to wait for the upcoming daemon to revoke it; but I think it's better to just allow revocation in this case, in case the daemon is running behind.

Reply:
Ah yup, good point.